language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/ExtractToFunctionCodeAction.java

```diff
@@ -0,0 +1,593 @@
+
+        if (enclosingModulePartNode.isEmpty()
+                || unSupportedModuleLevelSyntaxKinds.contains(enclosingModulePartNode.get().kind())) {
+            isRangeExtractable = false;
+        }
+
+        if (context.currentSyntaxTree().isEmpty() || (matchedCodeActionNode.kind() == SyntaxKind.MAPPING_CONSTRUCTOR
+                && matchedCodeActionNode.parent() != null
+                && matchedCodeActionNode.parent().kind() == SyntaxKind.TABLE_CONSTRUCTOR)) {
+            isRangeExtractable = false;
+        }
+
+        if (matchedCodeActionNode.kind() == SyntaxKind.FIELD_ACCESS
+                && (((FieldAccessExpressionNode) matchedCodeActionNode).expression().toSourceCode().strip()
+                .equals(SymbolUtil.SELF_KW)
+                || matchedCodeActionNode.parent().kind() == SyntaxKind.ASSIGNMENT_STATEMENT
+                || matchedCodeActionNode.parent().kind() == SyntaxKind.COMPOUND_ASSIGNMENT_STATEMENT)) {
+            isRangeExtractable = false;
+        }
+        return isRangeExtractable;
+    }
+
+    public static List<SyntaxKind> getSupportedExpressionSyntaxKindsList() {
+        return List.of(
+                SyntaxKind.BINARY_EXPRESSION,
+                SyntaxKind.BRACED_EXPRESSION,
+                SyntaxKind.QUALIFIED_NAME_REFERENCE,
+                SyntaxKind.INDEXED_EXPRESSION,
+                SyntaxKind.FIELD_ACCESS,
+                SyntaxKind.METHOD_CALL,
+                SyntaxKind.MAPPING_CONSTRUCTOR,
+                SyntaxKind.TYPEOF_EXPRESSION,
+                SyntaxKind.UNARY_EXPRESSION,
+                SyntaxKind.TYPE_TEST_EXPRESSION,
+                SyntaxKind.SIMPLE_NAME_REFERENCE,
+                SyntaxKind.LIST_CONSTRUCTOR,
+                SyntaxKind.TYPE_CAST_EXPRESSION,
+                SyntaxKind.TABLE_CONSTRUCTOR,
+                SyntaxKind.LET_EXPRESSION,
+                SyntaxKind.IMPLICIT_NEW_EXPRESSION,
+                SyntaxKind.EXPLICIT_NEW_EXPRESSION,
+                SyntaxKind.STRING_LITERAL,
+                SyntaxKind.NUMERIC_LITERAL,
+                SyntaxKind.BOOLEAN_LITERAL,
+                SyntaxKind.ERROR_CONSTRUCTOR
+        );
+    }
+
+    public static List<SyntaxKind> getSupportedStatementSyntaxKindsList() {
+        return List.of(
+                SyntaxKind.LIST, // not a statement
+                SyntaxKind.LOCAL_VAR_DECL,
+                SyntaxKind.ASSIGNMENT_STATEMENT,
+                SyntaxKind.IF_ELSE_STATEMENT,
+                SyntaxKind.WHILE_STATEMENT,
+                SyntaxKind.RETURN_STATEMENT,
+                SyntaxKind.COMPOUND_ASSIGNMENT_STATEMENT,
+                SyntaxKind.LOCK_STATEMENT,
+                SyntaxKind.FOREACH_STATEMENT,
+                SyntaxKind.MATCH_STATEMENT,
+                SyntaxKind.DO_STATEMENT
+        );
+    }
+
+    @Override
+    public String getName() {
+        return NAME;
+    }
+
+    static class ArgLists {
```

Comment:
Shall we rename this class to `ArgListsHolder`. And let's make the class and fields private